bootstore/src/schemes/v0/request_manager.rs

```diff
@@ -0,0 +1,401 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! A mechanism for tracking in flight requests
+
+use super::{Config, Envelope, Msg, Request, RequestType, Share, SharePkg};
+use sled_hardware::Baseboard;
+use std::collections::{BTreeMap, BTreeSet};
+use std::time::Instant;
+use uuid::Uuid;
+
+/// Acknowledgement tracking for `RequestType::InitRack`.
+#[derive(Debug, Default)]
+pub struct InitAcks {
+    pub expected: BTreeSet<Baseboard>,
+    pub received: BTreeSet<Baseboard>,
+}
+
+/// Acknowledgement tracking for `RequestType::LoadRackSecret` and
+/// `RequestType::Learn`
+#[derive(Debug)]
+pub struct ShareAcks {
+    pub threshold: u8,
+    pub received: BTreeMap<Baseboard, Share>,
+}
+
+impl ShareAcks {
+    pub fn new(threshold: u8) -> ShareAcks {
+        ShareAcks { threshold, received: BTreeMap::new() }
+    }
+}
+
+/// A mechanism to track in flight requests
+///
+/// Manages expiry, acknowledgment, and retries
+#[derive(Debug)]
+pub enum TrackableRequest {
+    /// A request from the caller of the Fsm API to initialize a rack
+    ///
+    /// This must only be called at one peer, exactly once. That peer
+    /// will be in `InitialMember` state.
+    InitRack {
+        rack_uuid: Uuid,
+        packages: BTreeMap<Baseboard, SharePkg>,
+        acks: InitAcks,
+    },
+
+    /// A request from the caller of the Fsm API to load a rack secret
+    ///
+    /// Only peers in `InitialMember` or `Learned` state can load rack secrets
+    LoadRackSecret { rack_uuid: Uuid, acks: ShareAcks },
+
+    /// A request received from a peer to learn a new share
+    ///
+    /// This request results in share gathering so "extra" shares can be
+    /// decrypted via a key derived from the rack secret and one of them
+    /// distributed to the learner.
+    ///
+    /// Only peers in `InitialMember` state can respond successfully
+    LearnReceived { rack_uuid: Uuid, from: Baseboard, acks: ShareAcks },
+
+    /// A request sent from a peer in `Learning` state to another peer
+    /// to learn a key share.
+    LearnSent { to: Baseboard },
+}
+
+/// A mechanism to manage all in flight requests
+///
+/// We expect very few requests at a time - on the order of one or two requests.
+pub struct RequestManager {
+    id: Baseboard,
+    config: Config,
+    requests: BTreeMap<Uuid, TrackableRequest>,
+    expiry_to_id: BTreeMap<Instant, Uuid>,
+
+    /// Messages that need sending to other peers.
+    ///
+    /// These should be drained on each API call.
+    envelopes: Vec<Envelope>,
```

Comment:
interesting that we have a `Vec<Envelope>` both here and in the parent `Fsm`; are both necessary?

Reply:
e.g. maybe we _shouldn't_ pass `self.envelopes` from the parent `Fsm` (see this comment https://github.com/oxidecomputer/omicron/pull/3385/files#r1256493304), but should instead _always_ drain the `RequestManager`'s `envelopes` into the parent?

Reply:
I'm not really sure what you are trying to get at with your latter comment. While it's possible both are not neccessary, I think it's the simplest solution. I also ensure both get drained efficiently with the `Fsm::drain_envelopes` call.

Additionally, I changed `Fsm::envelopes` to `Fsm::responses` as you suggested in a prior comment. So maybe they don't seem so redundant anymore given currrent naming 😆